File: 129.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root: return 0
        res = []
        self.dfs(root, [], res)
        res = [int(''.join(map(str, item))) for item in res]
        return sum(res)
    def dfs(self, root, path, res):
        if not root.left and not root.right:
            path.append(root.val)
            res.append(path)
            return
        if root.left:
            self.dfs(root.left, path + [root.val], res)
        if root.right:
            self.dfs(root.right, path + [root.val], res)

File: 129.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    int sumNumbers(TreeNode* root) {
        if (root == NULL){
            return res;
        }
        dfs(root, 0);
        return res;
    }
    void dfs(TreeNode* root, int num){
        if(root == NULL){
            return ;
        }
        num = 10 * num + root->val;
        if(!root->left and !root->right){
            res += num;
        }
        if(root->left){
            dfs(root->left, num);
        }
        if(root->right){
            dfs(root->right, num);
        }
    }
};
